k8s/src/crd.rs

```diff
@@ -0,0 +1,516 @@
+use k8s_openapi::api::core::v1::{
+    Container, EnvVar, Pod, PodSpec, ResourceRequirements, Service, ServicePort, ServiceSpec,
+    Volume, VolumeMount,
+};
+use kube::api::ObjectMeta;
+use kube::CustomResource;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+pub const DEFAULT_CUDA_IMAGE: &str = "persiaml/persia-cuda-runtime:latest";
+pub const DEFAULT_CPU_IMAGE: &str = "persiaml/persia-cpu-runtime:latest";
+
+pub fn get_emb_server_pod_name(job_name: &str, replica_index: usize) -> String {
+    format!("{}-embedding-server-{}", job_name, replica_index)
+}
+
+pub fn get_mid_server_pod_name(job_name: &str, replica_index: usize) -> String {
+    format!("{}-middware-server-{}", job_name, replica_index)
+}
+
+pub fn get_trainer_pod_name(job_name: &str, replica_index: usize) -> String {
+    format!("{}-trainer-{}", job_name, replica_index)
+}
+
+pub fn get_dataloader_pod_name(job_name: &str, replica_index: usize) -> String {
+    format!("{}-dataloader-{}", job_name, replica_index)
+}
+
+pub fn get_metrics_gateway_pod_name(job_name: &str) -> String {
+    format!("{}-metrics-gateway", job_name)
+}
+
+pub fn get_metrics_gateway_service_name(job_name: &str) -> String {
+    format!("{}-metrics-gateway-service", job_name)
+}
+
+pub fn get_label_selector(job_name: &str) -> String {
+    format!("persia_job={}", job_name)
+}
+
+#[derive(CustomResource, Serialize, Deserialize, Debug, PartialEq, Clone, JsonSchema)]
+#[kube(
+    group = "persia.com",
+    version = "v1",
+    kind = "PersiaJob",
+    plural = "persiajobs",
+    derive = "PartialEq",
+    namespaced
+)]
+#[allow(non_snake_case)]
```

Comment:
why non_snake_case? json_schema can't be parse?

Reply:
To be consistent with k8s, e.g. https://kubernetes.io/docs/concepts/storage/volumes/#aws-ebs-configuration-example